crates/driver/src/infra/tokens.rs

```diff
@@ -0,0 +1,88 @@
+use {
+    crate::{domain::eth, infra::Ethereum},
+    anyhow::Result,
+    contracts::ERC20,
+    ethcontract::dyns::DynWeb3,
+    std::{
+        collections::HashMap,
+        sync::{Arc, RwLock},
+    },
+};
+
+#[derive(Clone, Debug, Default)]
+pub struct Metadata {
+    pub decimals: u8,
+    pub symbol: String,
+}
+
+/// Provides metadata of tokens.
+pub struct Fetcher {
+    web3: DynWeb3,
+    cache: Arc<RwLock<HashMap<eth::TokenAddress, Metadata>>>,
+}
+
+impl Fetcher {
+    pub fn new(eth: &Ethereum) -> Self {
+        Self {
+            web3: eth.web3().clone(),
+            cache: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Fetches metadata of the requested token from a node.
+    async fn fetch_token_infos(
+        &self,
+        addresses: &[eth::TokenAddress],
+    ) -> Vec<Result<(eth::TokenAddress, Metadata)>> {
+        let futures = addresses.iter().map(|address| async {
+            let erc20 = ERC20::at(&self.web3, address.0 .0);
+            // Use `try_join` because these calls get batched under the hood
+            // so if one of them fails the other will as well.
+            // Also this way we won't get incomplete data for a token.
+            let (decimals, symbol) = futures::future::try_join(
+                erc20.methods().decimals().call(),
+                erc20.methods().symbol().call(),
```

Comment:
We didn't do this before - but I think that we should consider fetching balances as well here. Specifically, the existing token buffers allow solvers to make additional optimizations or shortcuts that might be useful to know at quoting time.

I **don't think it should be part of this PR** - just wanted to point it out as food for thought.

Reply:
> Specifically, the existing token buffers allow solvers to make additional optimizations or shortcuts that might be useful to know at quoting time.

I thought we explicitly wanted to forbid internalisation while quoting. Or do you have other optimisations in mind?